packages/core/test/Proxy.ts

```diff
@@ -0,0 +1,153 @@
+    stakingContract = await Staking.deploy(HMTokenContract.address, 1, 1);
+    await stakingContract.deployed();
+
+    const EscrowFactory = await ethers.getContractFactory('EscrowFactory');
+    escrowFactoryProxyContract = await upgrades.deployProxy(
+      EscrowFactory,
+      [HMTokenContract.address, stakingContract.address],
+      { initializer: 'initialize' }
+    );
+    await escrowFactoryProxyContract.deployed();
+
+    adminImplementation = await upgrades.admin.getInstance();
+  });
+
+  it('Deploy proxy factory', async function () {
+    const adminAddress = await upgrades.erc1967.getAdminAddress(
+      escrowFactoryProxyContract.address
+    );
+    const implementationAddress =
+      await upgrades.erc1967.getImplementationAddress(
+        escrowFactoryProxyContract.address
+      );
+
+    expect(adminAddress).to.equal(adminImplementation.address);
+    expect(adminAddress).to.equal(
+      await adminImplementation.getProxyAdmin(
+        escrowFactoryProxyContract.address
+      )
+    );
+    expect(implementationAddress).to.equal(
+      await adminImplementation.getProxyImplementation(
+        escrowFactoryProxyContract.address
+      )
+    );
+  });
+
+  it('Check proxy storage', async function () {
+    expect(HMTokenContract.address).to.equal(
+      await escrowFactoryProxyContract.eip20()
+    );
+    expect(stakingContract.address).to.equal(
+      await escrowFactoryProxyContract.staking()
+    );
+  });
+
+  it('Check implementation storage', async function () {
+    const escrowFactoryImplementation = await ethers.getContractAt(
+      'EscrowFactory',
+      await upgrades.erc1967.getImplementationAddress(
+        escrowFactoryProxyContract.address
+      )
+    );
+    expect(await escrowFactoryImplementation.eip20()).to.equal(
+      ethers.constants.Zero
+    );
+    expect(await escrowFactoryImplementation.staking()).to.equal(
+      ethers.constants.Zero
+    );
+  });
+
+  it('Upgrade proxy', async function () {
+    expect(
+      await escrowFactoryProxyContract.hasEscrow(owner.getAddress())
+    ).to.equal(false);
+
+    const EscrowFactory = await ethers.getContractFactory(
+      'EscrowFactoryUpgradeTest'
+    );
+    escrowFactoryProxyContract = await upgrades.upgradeProxy(
+      escrowFactoryProxyContract,
+      EscrowFactory
+    );
+
+    try {
+      escrowFactoryProxyContract.hasEscrow(owner.getAddress());
+    } catch (error: any) {
+      assert(
+        error.message ===
+          'escrowFactoryProxyContract.hasEscrow is not a function'
+      );
+    }
+
+    expect(HMTokenContract.address).to.equal(
+      await escrowFactoryProxyContract.eip20()
+    );
+    expect(stakingContract.address).to.equal(
+      await escrowFactoryProxyContract.staking()
+    );
+  });
+
+  it('Admin permissions', async function () {
+    expect(await adminImplementation.owner()).to.equal(
+      await owner.getAddress()
+    );
+
+    try {
```

Comment:
Use [revert asserts](https://hardhat.org/hardhat-chai-matchers/docs/overview#reverts) instead of try catch